Scenes/SongPlayer/song_player.gd

```diff
@@ -0,0 +1,89 @@
+extends Control
+
+enum LineMode{
+	SingleLine,
+	N_Line,
+	H_Scrolling,
+	V_Scrolling_N_Line,
+	V_Scrolling_unlimited,
+}
+enum CPUPlayerMode{
+	WPM,
+	SongSync,
+	Disabled,
+}
+
+@export var line_mode:LineMode = LineMode.SingleLine
+@export var cpu_player_mode:CPUPlayerMode = CPUPlayerMode.Disabled
+
+# text probably needs an annotation but i'm unsure which one would be most appropriate
+var text:String = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor\
+ incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation \
+ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in \
+voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non \
+proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
+var char_per_line:int = 50
+#var line_index:int = 0
+
+var player_cursor:int = 0
+var computer_cursor:int = 0
+
+# Called when the node enters the scene tree for the first time.
+func _ready() -> void:
+	sync_text_elements()
+	pass # Replace with function body.
+
+
+# Called every frame. 'delta' is the elapsed time since the previous frame.
+func _process(delta: float) -> void:
+	update_text_interface()
+	pass
+	
+
+func sync_text_elements() -> void:
+	pass #TODO format of all text elements needs to be identical or else display will not function properly.
+
+func update_text_interface() -> void:
+	#guard cases
+	if player_cursor>text.length() or computer_cursor>text.length():
+		return
+	#clear text
+	$TextInterface/TextDisplay.text = ""
+	$TextInterface/CPUText.text = ""
+	$TextInterface/PlayerText.text = ""
+	$TextInterface/SharedText.text = ""
```

Comment:
I think the code would be easier to understand if there was an explanation for what each Node gets used for:

```suggestion
	# This Node is for the white background text.
	$TextInterface/TextDisplay.text = ""
	# This Node is for the red text that has only been typed by the enemy so far.
	$TextInterface/CPUText.text = ""
	# This Node is for the green text that has only been typed by the player so far.
	$TextInterface/PlayerText.text = ""
	# This Node is for the blue text that has been typed by both the enemy and the player.
	$TextInterface/SharedText.text = ""
```

Reply:
It looks like you implemented this slightly differently. I like your implementation better.